### PR TITLE
[SPARK-34702][SQL] Avoid unnecessary code generation in JoinCodegenSupport.genBuildSideVars

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

As a followup from code review in https://github.com/apache/spark/pull/31736#discussion_r588134104 , for `JoinCodegenSupport.genBuildSideVars`, we only need to generate build side variables with default values for LEFT OUTER and RIGHT OUTER join, but not for other join types (i.e. LEFT SEMI and LEFT ANTI). Create this PR to clean up the code.

In addition, change `BroadcastNestedLoopJoinExec` unit test to cover both whole stage code-gen enabled and disabled. Harden the unit tests to exercise all code paths.

### Why are the changes needed?

Avoid unnecessary code generation.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing unit tests.
* BHJ and SHJ inner join is covered in `InnerJoinSuite.scala`
* BHJ and SHJ left outer and right outer join are covered in `OuterJoinSuite.scala`
* BHJ and SHJ left semi, left anti and existence join are covered in `ExistenceJoinSuite.scala`

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala [hunk 1]
```diff
@@ -46,7 +46,7 @@ case class BroadcastHashJoinExec(
     left: SparkPlan,
     right: SparkPlan,
     isNullAwareAntiJoin: Boolean = false)
-  extends HashJoin with CodegenSupport {
+  extends HashJoin {
```
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:49]: This is just a minor fix packed in this PR. `HashJoin` already extends `JoinCodegenSupport`. So we don't need `CodegenSupport` here.

**Comment by c21** [comment]
@cloud-fan and @maropu could you help take a look when you have time, thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:49]: If you don't mind, could you make a separate PR? You can use `[MINOR]` tag instead of creating a new JIRA.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/ExistenceJoinSuite.scala [hunk 2]
```diff
@@ -103,7 +103,7 @@ class ExistenceJoinSuite extends SparkPlanTest with SharedSparkSession {
       ProjectExec(output, FilterExec(condition, join))
     }
 
-    test(s"$testName using ShuffledHashJoin") {
+    testWithWholeStageCodegenOnAndOff(s"$testName using ShuffledHashJoin") { _ =>
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/ExistenceJoinSuite.scala:106]: This test enforcement looks orthogonal to this PR (`Avoiding unncessary code generation`). Can we do this as a test PR independently? We can merge it separately to the other branches.

**Comment by cloud-fan** [comment]
The code cleanup LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/ExistenceJoinSuite.scala (hunk 2 above)
Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/ExistenceJoinSuite.scala:106]: @dongjoon-hyun - if the purpose of separating PRs is to merge change to other branches as well, I can submit another PR for all test-related change, i.e. the change in `InnerJoinSuite.scala`, `OuterJoinSuite.scala` and `ExistenceJoinSuite.scala`.

Do we want to do that? These test changes are valid to be merged into `3.1` branch as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala (hunk 1 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:49]: @dongjoon-hyun - sure, will do it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/ExistenceJoinSuite.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/ExistenceJoinSuite.scala:106]: Yes, I want to merge the test improvement into branch-3.1.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/ExistenceJoinSuite.scala:106]: SPARK-34702 is for Apache Spark 3.2.0.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/InnerJoinSuite.scala [hunk 3]
```diff
@@ -198,7 +198,7 @@ class InnerJoinSuite extends SparkPlanTest with SharedSparkSession {
       }
     }
 
-    test(s"$testName using BroadcastNestedLoopJoin build left") {
+    testWithWholeStageCodegenOnAndOff(s"$testName using BroadcastNestedLoopJoin build left") { _ =>
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/InnerJoinSuite.scala:201]: `BroadcastNestedLoopJoin` codegen is merged in `3.2` (current master), so keep test change here.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/InnerJoinSuite.scala [hunk 4]
```diff
@@ -207,7 +207,7 @@ class InnerJoinSuite extends SparkPlanTest with SharedSparkSession {
       }
     }
 
-    test(s"$testName using BroadcastNestedLoopJoin build right") {
+    testWithWholeStageCodegenOnAndOff(s"$testName using BroadcastNestedLoopJoin build right") { _ =>
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/InnerJoinSuite.scala:210]: same reason as change above.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/ExistenceJoinSuite.scala (hunk 2 above)
Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/ExistenceJoinSuite.scala]: @dongjoon-hyun - sure, updated.

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @c21 . I also merged the spin-off test PR.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by c21** [comment]
Thank you @cloud-fan and @dongjoon-hyun for review!

**Comment by maropu** [comment]
late lgtm.

**Comment by c21** [comment]
Thank you @maropu for review as well!
